File: tools/viz_flavour.py
# viz_flavour.py — plot the dump from viz_flavour.c.
#
# Part of the Hex9 (H9) Project
# Copyright ©2026, Ben Griffin
# Licensed under the Apache License, Version 2.0
#
# Two panels over the same grid of encoded sample points:
#   left  — coloured by WALK BIN (h9_bin; byte-identical to Python = truth)
#   right — coloured by IDENTITY resolution (ring rendered from the full uuid)
# Black outlines: the disk(2) cell polygons (canonical kring emission).
# Where the right panel's colour regions cut THROUGH outlines, identity is
# resolving those points to the wrong cell — the mode-1-half defect.
#
# Usage: viz_flavour <dump.csv> <out.png> [title]

import sys
import csv
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

dump, out = sys.argv[1], sys.argv[2]
title = sys.argv[3] if len(sys.argv) > 3 else ""

pts = []            # (lon, lat, walkbin, ringhash)
rings = {}          # cellidx -> [(lon, lat), ...]
with open(dump) as f:
    for row in csv.reader(f):
        if row[0] == "P":
            pts.append((float(row[1]), float(row[2]), row[3], row[4]))
        elif row[0] == "R":
            rings.setdefault(int(row[1]), []).append((float(row[3]), float(row[4])))

def colour_map(keys):
    # one visually-distinct colour per key — collisions in a small cycled
    # palette previously made adjacent same-coloured cells read as split
    # hexagons (the phantom "striping")
    uniq = sorted(set(keys))
    cmap = plt.get_cmap("gist_ncar")
    n = max(len(uniq), 2)
    return {k: cmap(0.05 + 0.9 * i / (n - 1)) for i, k in enumerate(uniq)}

lons = [p[0] for p in pts]
lats = [p[1] for p in pts]
wcol = colour_map([p[2] for p in pts])
icol = colour_map([p[3] for p in pts])

fig, axes = plt.subplots(1, 2, figsize=(16, 8), sharex=True, sharey=True)
for ax, key, name in ((axes[0], 2, "walk bins (h9_bin == Python: truth)"),
                      (axes[1], 3, "identity resolution (ring of full uuid)")):
    ax.scatter(lons, lats, c=[(wcol if key == 2 else icol)[p[key]] for p in pts],
               s=4, marker="s", linewidths=0)
    for poly in rings.values():
        xs = [v[0] for v in poly]
        ys = [v[1] for v in poly]
        ax.plot(xs, ys, "k-", linewidth=1.2)
    # annotate each colour region with the trailing digits of its key at the
    # region's mean position (walk panel: last 3 body digits of the bin label)
    regions = {}
    for p in pts:
        regions.setdefault(p[key], []).append((p[0], p[1]))
    for k, ps in regions.items():
        if len(ps) < 12:
            continue
        mx = sum(v[0] for v in ps) / len(ps)
        my = sum(v[1] for v in ps) / len(ps)
        if key == 2:
            # bin hex: body nibbles up to sentinel + tail nibble
            body = k.split("f" * 6)[0]
            tag = body[-3:] + "." + k[-1]
        else:
            tag = k[-3:]
        ax.text(mx, my, tag, fontsize=6, ha="center", va="center")
    ax.set_title(name)
    ax.set_aspect("equal", adjustable="box")

fig.suptitle(title)
fig.tight_layout()
fig.savefig(out, dpi=140)
print(f"wrote {out}  ({len(pts)} samples, {len(rings)} outlines)")

File: tools/viz_flavour.c
/* viz_flavour.c — dump data to visualise the mode-1-half flavour resolution
 * defect (docs/addressing-doctrine.md, the wip/identity-reversible-walk
 * blocker).
 *
 * Part of the Hex9 (H9) Project
 * Copyright ©2026, Ben Griffin
 * Licensed under the Apache License, Version 2.0
 *
 * For an N×N grid of points covering a cell and its 2-ring at layer L,
 * encode each point and report:
 *   - its WALK BIN (h9_bin — byte-identical to the Python reference), and
 *   - its IDENTITY cell (the ring rendered from the full uuid, hashed),
 * plus the outlines of the disk(2) cells (rendered from kring's canonical
 * uuids) for overlay.
 *
 * Usage: viz_flavour lon lat L N [encoder]  > dump.csv
 *        encoder: 0 = nn/beam (default), 1 = containment
 * Lines:  P,lon,lat,walkbin_hex,ringhash      (grid samples)
 *         R,cellidx,vtx,lon,lat               (outline vertices)
 */
#include "hex9_c.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define RING_N 7

/* FNV-1a over the ring doubles — stable id for "which cell" */
static unsigned long ring_hash(const double *ring) {
    const unsigned char *b = (const unsigned char *)ring;
    unsigned long h = 1469598103934665603UL;
    for (size_t i = 0; i < 2 * RING_N * sizeof(double); i++) {
        h ^= b[i];
        h *= 1099511628211UL;
    }
    return h;
}

int main(int argc, char **argv) {
    if (argc < 5) { fprintf(stderr, "usage: viz_flavour lon lat L N\n"); return 2; }
    const double lon0 = atof(argv[1]), lat0 = atof(argv[2]);
    const int L = atoi(argv[3]), N = atoi(argv[4]);

    char err[256];
    if (hex9_warp_init(err, sizeof err)) { fprintf(stderr, "warp: %s\n", err); return 1; }
    if (argc > 5) hex9_set_encoder(atoi(argv[5]));

    uint8_t centre[16];
    if (hex9_encode(lon0, lat0, centre) != 0) { fprintf(stderr, "encode failed\n"); return 1; }

    /* disk(2) outlines from canonical kring emission */
    uint8_t disk2[19 * 16];
    const int64_t nd = hex9_k_disk(centre, L, 2, disk2, 19);
    if (nd <= 0) { fprintf(stderr, "k_disk failed\n"); return 1; }
    double bb_lon_min = 1e30, bb_lon_max = -1e30, bb_lat_min = 1e30, bb_lat_max = -1e30;
    for (int j = 0; j < (int)nd; j++) {
        double ring[2 * RING_N];
        if (hex9_cell_ring(disk2 + (size_t)j * 16, L, 0, ring, RING_N) != RING_N) continue;
        for (int k = 0; k < RING_N; k++) {
            printf("R,%d,%d,%.12f,%.12f\n", j, k, ring[2*k], ring[2*k + 1]);
            if (k < 6) {
                if (ring[2*k]   < bb_lon_min) bb_lon_min = ring[2*k];
                if (ring[2*k]   > bb_lon_max) bb_lon_max = ring[2*k];
                if (ring[2*k+1] < bb_lat_min) bb_lat_min = ring[2*k + 1];
                if (ring[2*k+1] > bb_lat_max) bb_lat_max = ring[2*k + 1];
            }
        }
    }

    /* sample grid over the disk(2) bbox (slightly shrunk to stay relevant) */
    const double pad = -0.02;
    const double dl = (bb_lon_max - bb_lon_min) * pad, dphi = (bb_lat_max - bb_lat_min) * pad;
    bb_lon_min -= dl; bb_lon_max += dl; bb_lat_min -= dphi; bb_lat_max += dphi;

    for (int iy = 0; iy < N; iy++) {
        for (int ix = 0; ix < N; ix++) {
            const double lon = bb_lon_min + (bb_lon_max - bb_lon_min) * (ix + 0.5) / N;
            const double lat = bb_lat_min + (bb_lat_max - bb_lat_min) * (iy + 0.5) / N;
            uint8_t full[16], bin[16];
            if (hex9_encode(lon, lat, full) != 0) continue;
            if (hex9_bin(full, L, bin) != 0) continue;
            double ring[2 * RING_N];
            if (hex9_cell_ring(full, L, 0, ring, RING_N) != RING_N) continue;
            char binhex[33];
            for (int b = 0; b < 16; b++) sprintf(binhex + 2*b, "%02x", bin[b]);
            printf("P,%.12f,%.12f,%s,%lu\n", lon, lat, binhex, ring_hash(ring));
        }
    }
    return 0;
}
